File: csharp/patterns/src/PrototypePatterns/Behavioral/ChainOfResponsibility/ChainOfResponsibilityPattern.cs
namespace PrototypePatterns.Behavioral.ChainOfResponsibility;

public abstract class Handler(Handler? next)
{
    private Handler? Next { get; } = next;

    public abstract string Handle(int level);

    protected string NextOrDefault(int level) => Next is null ? $"unhandled:{level}" : Next.Handle(level);
}

public sealed class TeamLeadHandler(Handler? next) : Handler(next)
{
    public override string Handle(int level) => level <= 1 ? "team-lead" : NextOrDefault(level);
}

public sealed class ManagerHandler(Handler? next) : Handler(next)
{
    public override string Handle(int level) => level <= 2 ? "manager" : NextOrDefault(level);
}

public sealed class DirectorHandler(Handler? next) : Handler(next)
{
    public override string Handle(int level) => level <= 3 ? "director" : NextOrDefault(level);
}

public static class ChainOfResponsibilityPattern
{
    public static IReadOnlyList<string> Demo()
    {
        Handler chain = new TeamLeadHandler(new ManagerHandler(new DirectorHandler(null)));
        return [chain.Handle(1), chain.Handle(2), chain.Handle(4)];
    }
}

File: java/patterns/src/main/java/org/prototypepatterns/behavioral/chainofresponsibility/ChainOfResponsibilityPattern.java
package org.prototypepatterns.behavioral.chainofresponsibility;

import java.util.List;

public final class ChainOfResponsibilityPattern {
    private ChainOfResponsibilityPattern() {
    }

    abstract static class Handler {
        private final Handler next;

        Handler(Handler next) {
            this.next = next;
        }

        abstract String handle(int level);

        String nextOrDefault(int level) {
            if (next == null) {
                return "unhandled:" + level;
            }
            return next.handle(level);
        }
    }

    static final class TeamLeadHandler extends Handler {
        TeamLeadHandler(Handler next) {
            super(next);
        }

        @Override
        String handle(int level) {
            if (level <= 1) {
                return "team-lead";
            }
            return nextOrDefault(level);
        }
    }

    static final class ManagerHandler extends Handler {
        ManagerHandler(Handler next) {
            super(next);
        }

        @Override
        String handle(int level) {
            if (level <= 2) {
                return "manager";
            }
            return nextOrDefault(level);
        }
    }

    static final class DirectorHandler extends Handler {
        DirectorHandler(Handler next) {
            super(next);
        }

        @Override
        String handle(int level) {
            if (level <= 3) {
                return "director";
            }
            return nextOrDefault(level);
        }
    }

    public static List<String> demo() {
        Handler chain = new TeamLeadHandler(new ManagerHandler(new DirectorHandler(null)));
        return List.of(chain.handle(1), chain.handle(2), chain.handle(4));
    }
}
